File: document/tutorial/calc/c/Interpreter.c
#include "Interpreter.h"
#include <stdio.h>

int interpret(Exp _p_)
{
  switch(_p_->kind)
  {
  case is_EAdd:
    return interpret(_p_->u.eadd_.exp_1) + interpret(_p_->u.eadd_.exp_2) ;
  case is_ESub:
    return interpret(_p_->u.eadd_.exp_1) - interpret(_p_->u.eadd_.exp_2) ;
  case is_EMul:
    return interpret(_p_->u.eadd_.exp_1) * interpret(_p_->u.eadd_.exp_2) ;
  case is_EDiv:
    return interpret(_p_->u.eadd_.exp_1) / interpret(_p_->u.eadd_.exp_2) ;
  case is_EInt:
    return _p_->u.eint_.integer_ ;
  default:
    fprintf(stderr, "Error: bad kind field when printing Exp!\n");
    exit(1);
  }
}

File: document/tutorial/calc/haskell/Interpreter.hs
module Interpreter where

import AbsCalc

interpret :: Exp -> Integer
interpret x = case x of
  EAdd exp0 exp  -> interpret exp0 + interpret exp
  ESub exp0 exp  -> interpret exp0 - interpret exp
  EMul exp0 exp  -> interpret exp0 * interpret exp
  EDiv exp0 exp  -> interpret exp0 `div` interpret exp
  EInt n  -> n
